lib/Transforms/ConvertToCiphertextSemantics/AssignLayout.cpp

```diff
@@ -396,12 +408,74 @@ Value implementUnpackOpForScalar(
   return splatOp.getResult();
 }
 
+std::string printRelation(const IntegerRelation &rel) {
+  std::string str;
+  llvm::raw_string_ostream os(str);
+  rel.print(os);
+  return str;
+}
+
+FailureOr<Value> implementAssignLayoutNew(
+    tensor_ext::AssignLayoutOp op, int64_t ciphertextSize,
+    ImplicitLocOpBuilder &builder,
+    const std::function<void(Operation *)> &createdOpCallback) {
+  NewLayoutAttr layout = dyn_cast<NewLayoutAttr>(op.getLayout());
+  if (!layout) {
+    return op.emitError()
+           << "Expected layout to be an IntegerRelation-style layout";
+  }
+
+  RankedTensorType dataSemanticType =
+      cast<RankedTensorType>(op.getValue().getType());
+  llvm::SmallVector<int64_t> dataSemanticShape;
+
+  IntegerRelation rel = layout.getIntegerRelation();
+  for (unsigned varPos = rel.getVarKindOffset(VarKind::Range);
+       varPos < rel.getNumRangeVars(); ++varPos) {
+    // The input variables are the data semantic tensor's axes.
+    std::optional<int64_t> dimBound =
+        rel.getConstantBound64(BoundType::UB, varPos);
+    if (!dimBound) {
+      return op.emitError()
+             << "Required all range variables to have an inferable upper "
+                "bound, "
+                "but found no upper bound for variable at position "
+             << varPos << " in relation " << printRelation(rel);
+    }
+    dataSemanticShape.push_back(dimBound.value());
+  }
+
+  // Last axis is always slot count in one ciphertext. This will need to be
+  // changed if/when we support hypercube slot structures.
```

Comment:
since the range (at least in the example) seems to be (ct, slot), then isn't the slot count already pushed into the shape above?

Reply:
True. Removed.